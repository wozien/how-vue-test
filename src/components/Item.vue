<template>
  <li>
    <a :href="item.url">{{ item.title }}</a>
    {{ item.url }}
  </li>
</template>

<script>
export default {
  props: ['item']
}
</script>

