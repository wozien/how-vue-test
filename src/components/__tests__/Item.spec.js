import { shallowMount, RouterLinkStub } from '@vue/test-utils'
import Item from '../Item.vue'
import merge from 'lodash.merge'


function createWrapper(mountConfig) {
  const config = {
    stubs: {
      RouterLink: RouterLinkStub
    },
    propsData: {
      item: {}
    }
  }
  return shallowMount(Item, merge(config, mountConfig))
}

describe('Item.vue', () => {
  test('renders item.url', () => {
    const item = {
      url: 'http://some-url.com',
    }

    const wrapper = createWrapper({
      propsData: {item}
    })
    expect(wrapper.text()).toContain('(some-url.com)')
  })

  test('renders item.url', () => {
    const item = {
      score: 10
    }

    const wrapper = createWrapper({
      propsData: {item}
    })
    expect(wrapper.text()).toContain('' + item.score)
  })

  test('renders item.by', () => {
    const item = {
      by: 10
    }

    const wrapper = shallowMount(Item, {
      propsData: { item }
    })
    expect(wrapper.text()).toContain('' + item.by)
  })

  test('renders a link to the item.url with item.title as test', () => {
    const item = {
      url: 'http://some-url.com',
      title: 'some-title'
    }

    const wrapper = createWrapper({
      propsData: {item}
    })
    const a = wrapper.find('a')
    expect(a.text()).toBe(item.title)
    expect(a.attributes().href).toBe(item.url)
  })

  test('renders the time since the last post', () => {
    const dateNow = jest.spyOn(Date, 'now')
    const dateNowTime = new Date('2021')

    dateNow.mockImplementation(() => dateNowTime)

    const item = {
      time: (dateNowTime / 1000) - 600
    }
    const wrapper = createWrapper({
      propsData: {item}
    })
    dateNow.mockRestore()
    expect(wrapper.text()).toContain('10 minutes ago')
  })

  // snapshot
  test('renders correctly', () => {
    const dateNow = jest.spyOn(Date, 'now')
    const dateNowTime = new Date('2021')

    dateNow.mockImplementation(() => dateNowTime)
    const item = {
      by: 'aaa',
      id: 1234,
      score: 10,
      time: (dateNowTime/1000) - 600,
      title: 'vue-test-utils',
      type: 'story',
      url: 'https://www.baidu.com'
    }
    const wrapper = createWrapper({
      propsData: { item }
    })
    expect(wrapper.element).toMatchSnapshot();
  })
})