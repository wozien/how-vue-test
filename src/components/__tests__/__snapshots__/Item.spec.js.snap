// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item.vue renders correctly 1`] = `
<li
  class="news-item"
>
  <span
    class="score"
  >
    10
  </span>
   
  <span
    class="title"
  >
    <a
      href="https://www.baidu.com"
      rel="noopener"
      target="_blank"
    >
      vue-test-utils
    </a>
     
    <span
      class="host"
    >
       (baidu.com)
    </span>
  </span>
   
  <br />
   
  <span
    class="meta"
  >
    <span
      class="by"
    >
      
      by 
      <a>
        aaa
      </a>
    </span>
     
    <span
      class="comments-link"
    >
      
      | 
      <a>
         comments
      </a>
    </span>
     
    <span>
      
      10 minutes ago
    
    </span>
  </span>
</li>
`;
